include/MySQL_HostGroups_Manager.h

```diff
@@ -390,6 +390,89 @@ class MySQL_HostGroups_Manager {
 #else
 	rwlock_t rwlock;
 #endif
+
+	enum HGM_TABLES {
+		MYSQL_SERVERS = 0,
+		MYSQL_REPLICATION_HOSTGROUPS,
+		MYSQL_GROUP_REPLICATION_HOSTGROUPS,
+		MYSQL_GALERA_HOSTGROUPS,
+		MYSQL_AWS_AURORA_HOSTGROUPS,
+		MYSQL_HOSTGROUP_ATTRIBUTES,
+
+		__HGM_TABLES_SIZE
+	};
+
+	std::array<uint64_t, __HGM_TABLES_SIZE> table_resultset_checksum;
+
+	class HostGroup_Server_Mapping {
+	public:
+		enum Type {
+			WRITER = 0,
+			READER = 1
+		};
+
+		struct Node {
+			MySrvC* srv = NULL;
+			unsigned int reader_hostgroup_id = -1;
+			unsigned int writer_hostgroup_id = -1;
+			MySerStatus server_status = MYSQL_SERVER_STATUS_OFFLINE_HARD;
+		};
+
+		HostGroup_Server_Mapping() : readonly_flag(1), myHGM(NULL) { }
+		~HostGroup_Server_Mapping() = default;
+
+		// Note: copy, remove, clear method also makes changes to MyHostGroups
+		void copy(Type dest_type, Type src_type, bool update_if_exists = true);
+		void remove(Type type, size_t index);
+		void clear(Type type);
+		//
+
+		inline
+		const std::vector<Node>& get(Type type) const {
+			return mapping[type];
+		}
+
+		inline
+		void add(Type type, Node& node) {
+			mapping[type].push_back(node);
+		}
+
+		inline
+		void set(Type type, const std::vector<Node>& nodes) {
+			mapping[type] = nodes;
+		}
+
+		inline
+		void set_readonly_flag(int val) {
+			readonly_flag = val;
+		}
+
+		inline
+		int get_readonly_flag() const {
+			return readonly_flag;
+		}
+
+		inline
+		void set_HGM(MySQL_HostGroups_Manager* hgm) {
+			myHGM = hgm;
+		}
+
+	private:
+		unsigned int get_hostgroup_id(Type type, size_t index) const;
+		unsigned int get_hostgroup_id(Type type, const Node& node) const;
+		MySrvC* insert_HGM(unsigned int hostgroup_id, const MySrvC* srv);
+		void remove_HGM(MySrvC* srv);
+
+		std::array<std::vector<Node>, 2> mapping;
```

Comment:
I don't like this particular type, I think it complicates the logic of multiples places in the class and doesn't payoff. The main use for having an array is using it for iteration, but there is only one place in which it's used for that.

Breaking this into two members, (e.g: `writer_mapping` and `reader_mapping`), would simplify a lot the flow for several functions and remove the necessity of helpers like `get_hostgroup_id`.

Reply:
Closing after already being discussed.